File: haskell/Pid.hs
{--
  PID control support for real and simulated flight controllers
 
  Copyright (C) 2024 Simon D. Levy
 
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, in version 3.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.
--} 

module Pid where

import Language.Copilot
import Copilot.Compile.C99

import Demands
import State
import Utils

type PidController = SBool -> SFloat -> State -> Demands -> Demands

 
pidController kp ki kd dt ilimit target actual capfun error' integ' = 
  (output, error, integ) where

  error = capfun $ target - actual

  integ = constrain (integ' + error * dt) (-ilimit) ilimit

  deriv = (error - error') / dt

  output = kp * error + ki * integ + kd * deriv


piController kp ki dt ilimit target actual integ' = (output, integ) where

  error = target - actual

  integ = constrain (integ' + error * dt) (-ilimit) ilimit

  output = kp * error + ki * integ

File: webots/controllers/python/pids/pid.py
import numpy as np


class PID:

    def __init__(self):

        self.error = 0
        self.integ = 0

    def _run(self, kp, ki, kd, ilimit, dt, target, actual, reset):

        error = target - actual

        deriv = (error - self.error) / dt

        self.error = 0 if reset else error

        self.integ = (0 if reset
                      else np.clip(self.integ + error * dt, -ilimit, +ilimit))

        return kp * error + kd * deriv

    def run_p(self, kp, dt, target, actual):

        return self._run(kp, 0, 0, 0, dt, target, actual, False)

    def run_pi(self, kp, ki, ilimit, dt, target, actual, reset=False):

        return self._run(kp, ki, 0, ilimit, dt, target, actual, reset)

    def run_pd(self, kp, kd, dt, target, actual, reset=False):

        return self._run(kp, 0, kd, 0, dt, target, actual, reset)
